Expand Testing section (#76)
tomjn/WordPress-The-Right-Way: testing/theory/tdd.md

# Test Driven Development

> Test Driven Development is a development process which takes the seemingly counter-intuitive approach of writing tests before writing the code that the tests cover.
> 
> The core idea of test driven development is to create failing tests that you are then able to compare your code against, which gives you assurance that your code at least performs the task that it was intended, based on the tests.
> 
> Test driven development has been around as long as there has been software development, but [Kent Beck](https://en.wikipedia.org/wiki/Kent_Beck) is credited with *rediscovering* it in the early 2000s.
> 

Reviewer: ```suggestion
Test Driven Development also known as TDD, is a development process which takes the seemingly counter-intuitive approach of writing tests before writing the code that the tests cover.

The core idea of test driven development is to create failing tests that you are then able to compare your code against, which gives you assurance that your code at least performs the task that it was intended, based on the tests.
```

> ## Steps of the TDD Process
> 
> 1. Add a test
> 1. Run all tests and see if the new test fails
> 1. Write the code
> 1. Run tests again to see if any tests fail
> 1. Refactor code
> 

Reviewer: ```suggestion
## Steps of the TDD Process

1. Add a test
2. Run all tests and see if the new test fails
3. Write the code
4. Run tests again to see if any tests fail
5. Refactor code

```

## Types of Testing

> As you can gather from the list above, the steps for TDD are quite simple, at a high level. But, much like with everything in software and web development, when things are simple, we are very good at complicating them. As a way of complicating matters, test driven development as a theory is agnostic to the specific type of testing that you perform. All it states is that the tests should be written first, and should be written specifically for the code that you are planning to write.

Reviewer: ```suggestion
As you can gather from the list above, the steps for TDD are quite simple, at a high level. Test driven development as a theory is agnostic to the specific type of testing that you perform. All it states is that the tests should be written first, and should be written specifically for the code that you are planning to write.
```

### Unit Testing

When getting started in TDD, it's almost uniform that you will be introduced to unit testing first. Unit testing is the core of most people's approach to TDD, as it takes the least amount of effort to setup. For PHP, you simply need to install PHPUnit, and start writing tests. There's more of a setup process for WordPress specifically, which you can read more about on the [unit testing](unit_testing.md) pages.

## Behavior Tests / Behavior Driven Development

Behavior driven development [(Behavior Driven Development)](bdd.md) is a testing methodology that emerged from TDD, with more of a focus on using plain language to write tests. When written out, BDD tests are generally easier to read as sentences than unit tests, or integration tests.

> 
> ### Integration Testing
> 
> _Need more information here_

Reviewer: ```suggestion

BDD tools can sometimes be used to automate integration testing, but are no substitute for a thorough manual QA check.
```
